cmd/bnsd/app/codec.proto

```diff
@@ -55,36 +57,102 @@ message Tx {
+      // No recursive batches!
+      nft.AddApprovalMsg add_approval_msg = 61;
+      nft.RemoveApprovalMsg remove_approval_msg = 62;
+      username.IssueTokenMsg issue_username_nft_msg = 63;
+      username.AddChainAddressMsg add_username_address_nft_msg = 64;
+      username.RemoveChainAddressMsg remove_username_address_msg = 65;
+      distribution.NewRevenueMsg new_revenue_msg = 66;
+      distribution.DistributeMsg distribute_msg = 67;
+      distribution.ResetRevenueMsg reset_revenue_msg = 68;
+      // upgrade schema is important enough, it should be a solo action
+      // aswap and gov don't make much sense as part of a batch (no vote buying)
+
+    }
+  }
+  repeated Union messages = 1 [(gogoproto.nullable) = false];
+}
+
+// ProposalOptions are possible items that can be enacted by a governance vote
+// Trimmed down somewhat arbitrary to what is believed to be reasonable
+message ProposalOptions {
+  oneof option {
+    cash.SendMsg send_msg = 51;
+    escrow.ReleaseEscrowMsg release_escrow_msg = 53;
+    escrow.UpdateEscrowPartiesMsg update_escrow_msg = 55;
+    validators.SetValidatorsMsg set_validators_msg = 58;
+    ProposalBatchMsg batch_msg = 60;
+    distribution.NewRevenueMsg new_revenue_msg = 66;
+    distribution.DistributeMsg distribute_msg = 67;
+    distribution.ResetRevenueMsg reset_revenue_msg = 68;
+    migration.UpgradeSchemaMsg upgrade_schema_msg = 69;
+    gov.UpdateElectorateMsg update_electorate_msg = 77;
+    gov.UpdateElectionRuleMsg update_election_rule_msg = 78;
+    gov.TextResolutionMsg text_resolution_msg = 79;
+  }
+}
+
+message ProposalBatchMsg {
+  message Union {
+    oneof sum {
+      cash.SendMsg send_msg = 51;
+      escrow.ReleaseEscrowMsg release_escrow_msg = 53;
+      escrow.UpdateEscrowPartiesMsg update_escrow_msg = 55;
+      validators.SetValidatorsMsg set_validators_msg = 58;
+      // no recursive batches
+      distribution.NewRevenueMsg new_revenue_msg = 66;
+      distribution.DistributeMsg distribute_msg = 67;
+      distribution.ResetRevenueMsg reset_revenue_msg = 68;
+      // don't allow UpgradeSchema as part of a batch, as effects are too confusing
+      gov.UpdateElectorateMsg update_electorate_msg = 77;
+      gov.UpdateElectionRuleMsg update_election_rule_msg = 78;
+      gov.TextResolutionMsg text_resolution_msg = 79;
+    }
+  }
+  repeated Union messages = 1 [(gogoproto.nullable) = false];
+}
+
+// AppCreateProposalMsg is byte compatible with CreateProposalMsg
+// assuming we use decodeProposalOptions.
 //
-//    repeated Union messages = 1 [(gogoproto.nullable) = false];
-//}
+// It is here to demonstrate a clean app-level usage of proposals.
+message AppCreateProposalMsg {
+  weave.Metadata metadata = 1;
+  gov.CreateProposalMsgBase base = 2;
+  ProposalOptions options = 3;
+}
+
+// AppProposal is byte compatible with Proposal
+// assuming we use decodeProposalOptions.
+//
+// It is here to demonstrate a clean app-level usage of proposals.
+message AppProposal {
```

Comment:
Q: Is this used?

Reply:
This is only used in test code, intended for clients, to easily create/parse options. I demo'd this in `x/demo` and it seemed a nice idea to folks. I can remove, no problem, if you wish

Reply:
I don't have the full picture. I trust you with this. 
Just as a side note: the same definition is in: https://github.com/iov-one/weave/blob/master/x/gov/sample_test.proto#L26
and `cmd/bnsd/app/AppProposal ` is not used (yet).

Reply:
Yes, you are correct. That was not used, I figured it would be a helper for eg. scenario tests. But then looking at the scenario tests you wrote, I realized this is not useful at all. Since we need the RawOptions in the Tx definition for server-side parsing. And if we build a tx client side, we will also need the same Tx. Or make a whole different Tx that includes these as embedded field rather than pre-marshalled...

I thought it was a neat trick to avoid manually marshalling stuff and make a cleaner interface. But I guess it didn't actually turn out to be useful. I could move all the Common/Base stuff from x/gov top-level again as well.

*sigh* wasted work, but better now than later.

Reply:
I will address that in a separate PR, as I do need to clean up some more cases for #613 as well